Update RenameTracking to make it behave like a fix for a diagnostic.

balajikris opened the pull request:
This change adds the following functionalities:
1. Show a preview of the changes to the current file in the preview pane.
2. Move the option to show preview changes to the preview pane.
3. We now also start the rename task during preview and reuse its result during the actual commit of the fix. This should bring some time savings due to starting the rename task early. Thanks to @dpoeschl for making this suggestion.

What is not yet done:
1. Make lightbulb on hover work.

This requires changes from platform and will be addressed in the near future. This is also related to another internal bug on @shyamnamboodiripad  to make lightbulb show on hover for suggestion fixes like remove unnecessary cast or remove 'this' qualification where there are no error squiggles.

*Note*: I am not looking to merge this PR for another couple of weeks. Currently computing information and displaying on preview pane is being done as a synchronous operation. An unfortunate side effect of this change will be that ctrl + . will appear unresponsive if  "RenameTracking" is a valid fix that is offered, since it will be the first fix offered and computing preview for this is an expensive task. But, the good part is the platform team is prototyping asynchronous preview pane and are on track to merge their changes soon. So, I will simply wait until we have their changes before I merge this PR. Until then, I'm looking to take feedback/modifications and prepare a commit to eventually push.

dpoeschl commented:
When the lightbulb item is committed, do they "cancel" the preview operation, causing us to lose our work if it's in flight?

It'd be nice to really be able to continue our work from first LB item preview until we're either committed or the lightbulb goes away. Maybe we could listen to diagnostics changed events or something? @heejaechang - Any suggestions of how we can continue our work between preview/commit when the final answer isn't yet computed, but without ignoring cancellation?

heejaechang commented:
@dpoeschl I think rename should just do rename current file and disable showing any errors. there is no point doing it for whole solution.

and I don't think we should let preview keep going when a user moves away from an item.

shyamnamboodiripad commented:
I haven't looked closely at the code - but are you assuming that the preview will always be kicked off before the actual fix is commited? Are you checking to avoid races (for instance both preview and actuial fix could get kicked off in cases where user presses ctrl+.+enter quickly and I wonder if that would introduce races in your code).

In general, I don't think we have a precedent for a fix that reuses computation that happens during preview :) I am not sure whether this is a good idea in general from API perspective. API assumes that preview operation and the actual fix are independent operations (or is this just my personal assumption? :)). I am not sure its a good idea to introduce dependency between preview and actual fix even if it works in the current implementation - I am not sure we would support this if we make changes in the future... :)

@dpoeschl regarding your question, I feel it is perfectly legal for the light bulb to cancel the preview operation when the fix is committed (since the preview UI will also be dismissed at this point). Trying to ignore cancellation could have undesirable side effects. For example, light bulb would also cancel the preview operation if user moved to the next fix in the light bulb menu (down arrow). Should we then allow the work to continue on and peg the user's CPU for an operation that user is not interested in)?

@Srivatsn, @heejaechang  - Thoughts?

balajikris commented:
Ah, forgot to update tests to look for 1 codeaction instead of 2. Will fix.

balajikris commented:
@heejaechang I'm not sure if there is a way to rename only a single file. What if you are renaming a reference and its definition is in some other file? also how'd we reason about conflicts and resolutions? @dpoeschl had more similar reasons as to why this has to be a full rename. 

@shyamnamboodiripad Yeah, the change addresses all the concerns mentioned in para 1 of your comment. No, it does not assume Preview is always invoked before Commit. There isn't any race introduced to the best of my knowledge. During Commit it waits on the rename task if it isn't already completed and kicks off a new one if the previous task wasn't initiated or if it didn't get to complete successfully.

Also, appreciate any guidance on a scenario such as this, where we have a potentially long running operation and we would like to avoid doing the work twice for preview and actual commit.

dpoeschl commented:
Addressing a few conversation threads:

Renaming only in current file
    - If we can find a way to rename only in the current file and get accurate complexification, errors, etc. in a fast way then I'd be okay with that. But if we can't get those things, then we either need a disclaimer ("This preview is not an accurate representation of the effect of your rename") or we need to get rid of the preview entirely. I don't want to set a precedent of hiding errors or showing fake code in the preview, because then you can never trust the preview as a real view into the future.

CodeActions reusing computation between preview & application
    - If we store the task on our custom RenameTrackingCodeAction (which we're doing), obey cancellation appropriately, and manage the dependence correctly, then I feel like we should be able to reuse that computation. As @balajikris points out, we're allowing the application step to use the preview step's calculation if it was completed before it got cancelled, but otherwise it just does the work itself. I don't think anything can fundamentally change in the universe between preview and apply phases because then you'd retrigger and we'd provide a different RenameTrackingCodeAction, right?

Obeying cancellation of the preview operations
    - I don't mind (too much) the cancellation of our work when the user moves between lightbulb items, but it's painful in situations where the rename process takes 6 seconds and the user waits 5 seconds for the preview before committing and we have to start the 6 seconds over again. But you're right @shyamnamboodiripad, the system definitely doesn't seem set up to support this (and Balaji's change doesn't include this behavior), and I think we'll be okay without it. We can always track and see how long the commits take, how long is spent in previews before committing, etc. before deciding to build out some larger system here.

heejaechang commented:
@balajikris there is no reason preview needs to perfect. if it is cheap sure, why not. but if it is expansive or require changing that doesn't support snapshot or forking, it should just skip it. that is why we support preview and invoke in two different path.

heejaechang commented:
@dpoeschl if preview takes more than 2 seconds to create preview, I am not sure why preview is even needed. it is preview, since I can quickly see what it would look like (again, doesn't need to be perfect). I think your next desire comes because it is trying to show something that shouldn't be in preview in the first place, and from there, everything becomes complicated. and breaking all simplicity LB design has to get around root issues.

shyamnamboodiripad commented:
I think it's fine to show an 'incomplete' preview of the future. The preview pane has limited real estate and I think its ok if the preview doesn't display the future in full fidelity :) However, I agree we should probably avoid showing an 'incorrect' preview. 

In general, I think the guidance has been that the preview operation should be quick to compute (so that user can get an idea of what the future is likely to look like without actually waiting for as long as it takes to apply the fix). Long running fixes should either provide their own 'cheap' preview if possible / skip the preview entirely.

In a large majority of cases that users use the rename tracking fix, I think the preview would be quick to calculate and accurate. And so I think there is value in displaying a preview for this fix. Yes there will be complicated cases where there are conflicts etc. where we won't be able to calculate a correct preview cheaply. Is it ok to skip displaying the preview in such cases (i.e. timeout after a fixed time interval with some indication as to why we don't have a preview)?

Btw, I am also not averse to checking the change as it is currently implemented and trying it out to see how it feels in dogfooding. The case where the preview takes long time to compute may be rare in practice and we can tweak the implementation based on dogfooding user experience.

heejaechang commented:
I agree with @shyamnamboodiripad if preview takes long time, I won't wait for it or even worse, I will probably think preview is broken. I am not sure whether that is any better than showing incorrect preview.

about timeout. I think timeout works if 90% of time it takes less than the timeout. otherwise, I think it is worse than either not showing preview at all from the start or showing incorrect preview.

DustinCampbell commented:
I just wanted to pop my head in here to clarify a design principle. This has been our design principle since the beginning when we created on light bulbs with live code previews years ago.

It is *not* OK to have a preview that shows an incorrect or broken picture of what the fix or refactoring will do. The purpose of the preview is (and always has been) to build the user's trust with the light bulb and show them what will happen to their code when they commit. If we show something that's incorrect, or even incomplete without an indicator, we will erode that trust. In such cases, it would be better to not show a preview at all rather than lie to the user with an approximation.

heejaechang commented:
Rename can choose to not show preview based on their own principle. But that's code fix implementer's decision not engine. API wise we certainly let user to show cheaper but not 100% correct future. - I believe we have work item to show some kind of indicator for this case in preview pane.

Also realistically, showing 100% accurate future only possible if all components code fix touches support snapshot and forking. if it requires any component that doesn't support them, ex, nuget package installation or change in vs state then there is no way to support it.

Personally, not showing at all or always showing 100% correct preview sounds like two extreme sides.



Sent from my Windows Phone
________________________________
From: Dustin Campbell<mailto:notifications@github.com>
Sent: ý3/ý2/ý2015 10:52 AM
To: dotnet/roslyn<mailto:roslyn@noreply.github.com>
Cc: HeeJae Chang<mailto:Heejae.Chang@microsoft.com>
Subject: Re: [roslyn] Update RenameTracking to make it behave like a fix for a diagnostic. (#908)


I just wanted to pop my head in here to clarify a design principle. This has been our design principle since the beginning when we created on light bulbs with live code previews years ago.

It is not OK to have a preview that shows an incorrect or broken picture of what the fix or refactoring will do. The purpose of the preview is (and always has been) to build the user's trust with the light bulb and show them what will happen to their code when they commit. If we show something that's incorrect, or even incomplete without an indicator, we will erode that trust. In such cases, it would be better to not show a preview at all rather than lie to the user with an approximation.

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/roslyn/pull/908#issuecomment-76778983>.

balajikris commented:
Thanks for the design feedback guys. I'm wondering if it is reasonable to check this in (after the async preview is in) and improve/tweak the implementation based on dogfooding feedback? Tagging @dpoeschl @rchande @heejaechang @shyamnamboodiripad @Pilchie for code review feedback..

heejaechang commented on src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingCodeAction.cs:
```diff
@@ -20,57 +20,62 @@ private class RenameTrackingCodeAction : CodeAction
+                return Task.FromResult(SpecializedCollections.SingletonEnumerable(committerOperation as CodeActionOperation));
+            }
+
+            protected override async Task<IEnumerable<CodeActionOperation>> ComputePreviewOperationsAsync(CancellationToken cancellationToken)
```
if rename changes xaml file as well, what happen?

heejaechang commented on src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingCodeAction.cs:
```diff
@@ -20,57 +20,62 @@ private class RenameTrackingCodeAction : CodeAction
                         }
 
-                        Debug.Assert(false, "RenameTracking codefix invoked on a document with a StateMachine which cannot invoke rename.");
+                        var snapshotSpan = stateMachine.TrackingSession.TrackingSpan.GetSpan(stateMachine.Buffer.CurrentSnapshot);
```
this seems can have a race. it starts with a snapshot (document) but from there it gets text buffer (mutable) and buffer.CurrentSnapshot which has no guarantee that current snapshot will be same as "document" snapshot.

how is it guarantee session data it gathered is for the "document" snapshot?

heejaechang commented on src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingCommitter.cs:
```diff
@@ -26,30 +27,30 @@ private class RenameTrackingCommitter : ForegroundThreadAffinitizedObject
+
+            // This task performs the actual rename on solution. We kick this off during preview to be able to show the diff preview
+            // and if it was successfully completed without being cancelled, we reuse its result during the actual commit operation as well.
+            private Task<RenameTrackingSolutionSet> _renameSymbolTask;
```
oh well, this is certainly not the way we imagined how people should use this API.

heejaechang commented on src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingCommitter.cs:
```diff
@@ -60,10 +61,48 @@ public void Commit(CancellationToken cancellationToken)
+                    cancellationToken,
+                    TaskScheduler.Default);
+
+                return await _renameSymbolTask.ConfigureAwait(false);
```
just return _renameSymbolTask ?

heejaechang replied:
hmm.. if you want to re-use _renameSymbolTask between preview and invoke, I think you shouldn't capture cancellationtoken, since preview and invoke has 2 different cancellationtoken.

heejaechang commented on src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingCommitter.cs:
```diff
@@ -81,59 +120,20 @@ private void RenameSymbolAndApplyChanges(CancellationToken cancellationToken, ou
+                else
                 {
-                    Contract.Fail("Invoked rename tracking smart tag but cannot find the symbol");
+                    renameTrackingSolutionSet = _renameSymbolTask.WaitAndGetResult(cancellationToken);
```
if you are going to do this, you can't capture cancellation for _renameSymbolTask.

VladimirReshetnikov commented on src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingCommitter.cs:
```diff
@@ -203,21 +202,21 @@ private Solution CreateSolutionWithOriginalName(Document document, CancellationT
                 symbol = symbol ?? semanticModel.GetSymbolInfo(token, cancellationToken).Symbol;
 
-                return symbol != null;
+                return symbol;
```
Consider merging the last 3 statements into a single 'return' statement.

VladimirReshetnikov replied:
Consider merging the last 3 lines to `return semanticFacts.GetDeclaredSymbol(semanticModel, token, cancellationToken) ?? semanticModel.GetSymbolInfo(token, cancellationToken).Symbol` (with a line break if necessary).

VladimirReshetnikov commented on src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingSolutionSet.cs:
```diff
@@ -0,0 +1,31 @@
+        /// </summary>
+        private class RenameTrackingSolutionSet
+        {
+            private readonly ISymbol _symbolToRename;
```
Consider removing the fields and using read-only auto-properties instead.

VladimirReshetnikov commented on src/EditorFeatures/Test/RenameTracking/RenameTrackingTestState.cs:
```diff
@@ -170,15 +170,14 @@ public void AssertTag(string expectedFromName, string expectedToName, bool invok
+            Assert.Equal(1, actions.Count);
 
             Assert.Equal(string.Format("Rename '{0}' to '{1}'", expectedFromName, expectedToName), actions[0].Title);
-            Assert.Equal(string.Format("Rename '{0}' to '{1}' with preview...", expectedFromName, expectedToName), actions[1].Title);
```
Cosinder using interpolated strings instead.

VladimirReshetnikov commented on src/EditorFeatures/Core/EditorFeatures.csproj:
```diff
@@ -599,6 +599,7 @@
     <Compile Include="Implementation\RenameTracking\RenameTrackingTaggerProvider.cs" />
```
Consider adding `<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->` at the top of the file after the xml declaration.

VladimirReshetnikov commented on src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingCommitter.cs:
```diff
@@ -60,10 +61,48 @@ public void Commit(CancellationToken cancellationToken)
+                var symbol = await TryGetSymbolAsync(solutionWithOriginalName, document.Id, cancellationToken).ConfigureAwait(false);
+                if (symbol == null)
+                {
+                    Contract.Fail("Invoked rename tracking smart tag but cannot find the symbol");
```
Nit: use trailing dots consistently (compare to line 81).

balajikris commented:
This is merged as #2248 . Closing this original PR.
